examples/language/lib/patterns/pattern_types.dart

```diff
@@ -0,0 +1,262 @@
+      case (var a, var b) && (var b, var c): // ...
+    }
+    // #enddocregion logical-and
+  }
+
+  {
+    var color = Color();
+    // #docregion logical-or
+    var isPrimary = switch (color) {
+      Color.red || Color.yellow || Color.blue => true,
+      _ => false
+    };
+    // #enddocregion logical-or
+  }
+
+  {
+    var shape = Square(10);
+    // #docregion share-guard
+    switch (shape) {
+      case Square(size: var s) || Circle(size: var s) when s > 0:
+        print('Non-empty symmetric shape');
+    }
+    // #enddocregion share-guard
+  }
+
+  {
+    // #docregion null-assert-match
+    List<String?> row = ['user', null];
+    switch (row) {
+      case ['user', var name!]: // ...
+      // 'name' is a non-nullable string here.
+    }
+    // #enddocregion null-assert-match
+  }
+
+  {
+    // #docregion null-assert-dec
+    (int?, int?) position = (2, 3);
+
+    var (x!, y!) = position;
+    // #enddocregion null-assert-dec
+  }
+
+  {
+    // #docregion null-check
+    String? maybeString = 'nullable with base type String';
+    switch (maybeString) {
+      case var s?:
+      // 's' has type non-nullable String here.
+    }
+    // #enddocregion null-check
+  }
+
+  {
+    // #docregion object
+    switch (shape) {
+      // Matches if shape is of type Rect, and then against the properties of Rect.
+      case Rect(width: var w, height: var h): // ...
+    }
+    // #enddocregion object
+  }
+
+  {
+    // #docregion object-getter
+    // Binds new variables x and y to the values of Point's x and y properties.
+    var Point(:x, :y) = Point(1, 2);
+    // #enddocregion object-getter
+  }
+
+  {
+    // #docregion record
+    var (myString: foo, myNumber: bar) = (myString: 'string', myNumber: 1);
+    // #enddocregion record
+  }
+
+  {
+    var record = (untyped: '1', typed: 2);
+    // #docregion record-getter
+    // Record pattern with variable subpatterns:
+    var (untyped: untyped, typed: int typed) = record;
+    var (:untyped, :int typed) = record;
+
+    switch (record) {
+      case (untyped: var untyped, typed: int typed): // ...
+      case (:var untyped, :int typed): // ...
+    }
+
+    // Record pattern wih null-check and null-assert subpatterns:
+    switch (record) {
+      case (checked: var checked?, asserted: var asserted!): // ...
+      case (:var checked?, :var asserted!): // ...
+    }
+
+    // Record pattern wih cast subpattern:
+    var (untyped: untyped as int, typed: typed as String) = record;
+    var (:untyped as int, :typed as String) = record;
+    // #enddocregion record-getter
+  }
+
+  {
+    // #docregion relational
+    String asciiCharType(int char) {
+      const space = 32;
+      const zero = 48;
+      const nine = 57;
+
+      return switch (char) {
+        < space => 'control',
+        == space => 'space',
+        > space && < zero => 'punctuation',
+        >= zero && <= nine => 'digit',
+        _ => ''
+      };
+    }
+    // #enddocregion relational
+  }
```

Comment:
We should eventually add tests for cases like these, but for now, let's add an assert to make the function used.

```suggestion
    assert(asciiCharType(32) == 'space');
  }
```